src/mono/mono/metadata/class-init.c

```diff
@@ -4197,65 +4197,76 @@ mono_class_set_runtime_vtable (MonoClass *klass, MonoVTable *vtable)
 }
 
 static int
-index_of_class (MonoClass *needle, MonoClass **haystack, int haystack_size) {
+index_of_class (MonoClass *needle, MonoVarianceSearchTableEntry *haystack, int haystack_size) {
 	for (int i = 0; i < haystack_size; i++)
-		if (haystack[i] == needle)
+		if (haystack[i].klass == needle)
 			return i;
 
 	return -1;
 }
 
 static void
-build_variance_search_table_inner (MonoClass *klass, MonoClass **buf, int buf_size, int *buf_count) {
-	if (!m_class_is_interfaces_inited (klass)) {
+build_variance_search_table_inner (MonoClass *klass, MonoVarianceSearchTableEntry *buf, int buf_size, int *buf_count, MonoClass *current) {
+	if (!m_class_is_interfaces_inited (current)) {
 		ERROR_DECL (error);
-		mono_class_setup_interfaces (klass, error);
+		mono_class_setup_interfaces (current, error);
 		return_if_nok (error);
 	}
-	guint c = m_class_get_interface_count (klass);
+
+	guint c = m_class_get_interface_count (current);
 	if (c) {
-		MonoClass **ifaces = m_class_get_interfaces (klass);
+		MonoClass **ifaces = m_class_get_interfaces (current);
 		for (guint i = 0; i < c; i++) {
 			MonoClass *iface = ifaces [i];
 			// Avoid adding duplicates or recursing into them.
 			if (index_of_class (iface, buf, *buf_count) >= 0)
 				continue;
 
 			if (mono_class_has_variant_generic_params (iface)) {
-				g_assert (*buf_count < buf_size);
-				buf[*buf_count] = iface;
+				int index = *buf_count;
 				(*buf_count) += 1;
+				g_assert (index < buf_size);
+				buf[index].klass = iface;
+				buf[index].depth = current->idepth;
+				buf[index].offset = mono_class_interface_offset (klass, iface);
 			}
 
-			build_variance_search_table_inner (iface, buf, buf_size, buf_count);
+			build_variance_search_table_inner (klass, buf, buf_size, buf_count, iface);
 		}
 	}
+
+	if (current->parent)
+		build_variance_search_table_inner (klass, buf, buf_size, buf_count, current->parent);
```

Comment:
probably doesn't matter, but we could avoid recursion in this case and just do a while loop that calls `build_variant_search_table_inner`